File: Day<27>/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    int catchThieves(char arr[], int n, int k)
    {
    	int result = 0;
    	vector<int> thief;
    	vector<int> police;

    	for (int i = 0; i < n; i++)
		{
    		if (arr[i] == 'P')
    			police.push_back(i);
    		else if (arr[i] == 'T')
    			thief.push_back(i);
    	}
    
    	// track lowest current indices of thief: thi[l], police: pol[r]
    	int l = 0, r = 0;
    	while (l < thief.size() && r < police.size())
		{
    		// can be caught
    		if (abs(thief[l] - police[r]) <= k)
			{
    			result++;
    			l++;
    			r++;
    		}
    		// increment the minimum index
    		else if (thief[l] < police[r])
    			l++;
    		else
    			r++;
    	}
    	return result;
    }
};

int main()
{
	int t;
	cin>>t;
	while (t--)
	{
		int n, k;
		cin>>n >> k;
		char arr[n];
		for (int i = 0; i < n; ++i)
			cin>>arr[i];
		Solution obj;
		cout<<obj.catchThieves(arr, n, k) << endl;
	}
	return 0;
}

File: Day<27>/Solution.py
class Solution:
    def catchThieves(self, arr, n, k):
    	i = 0
    	l = 0
    	r = 0
    	result = 0
    	thief = []
    	police = []

    	while i < n:
    		if arr[i] == 'P':
    			police.append(i)
    		elif arr[i] == 'T':
    			thief.append(i)
    		i += 1

    	# track lowest current indices of thief: thief[l], police: police[r]
    	while l < len(thief) and r < len(police):
    		
    		# can be caught
    		if (abs( thief[l] - police[r] ) <= k):
    			result += 1
    			l += 1
    			r += 1
    		
    		# increment the minimum index
    		elif thief[l] < police[r]:
    			l += 1
    		else:
    			r += 1
    	return result

if __name__=='__main__':
	t = int(input())
	for _ in range(t):
		line1 = list(map(int, input().strip().split()))
		n = line1[0]
		k = line1[1]
		arr = list(input().strip().split())
		obj = Solution()
		print(obj.catchThieves(arr, n, k))
